### PR TITLE
Symmetry: debug and new features

### PR DESCRIPTION
**Pull request by maki49**
## Fix some bugs and problems
- too many IBZ k-points     deepmodeling/abacus-develop#1606
- wrong bravais lattice and point group     deepmodeling/abacus-develop#1418
  * fix bugs with translation vector, euler angles, atom sort, etc.
  * optimize the algorithm for atom sort and finding translation vectors
## Features
- `rhog_symmetry`: symmetrizing charge density in reciprocal space
- space group analysis   deepmodeling/abacus-develop#1001
- `pricell`: primitive cell analysis     deepmodeling/abacus-develop#1387
- `gmatrix_convert` and `gtrans_convert`: converting a matrix / a vector from one lattice-vector-set to another
- `get_shortest_latvec`: finding the shortest lattice vectors from input lattice
## Unit test
- add unit tests for rotation matrices and translation vectors on different basis 
- add unit tests for new atom-ordering algorithm

### CODE REVIEW

**Comment by maki49** [comment]
I'm confused about why case `121_PW_KPAR` timed out ... I can't reproduce this problem on my workstation. @caic99

**Comment by caic99** [comment]
> I'm confused about why case `121_PW_KPAR` timed out ... I can't reproduce this problem on my workstation. @caic99

Hi @maki49 ,
I recommend to use a docker image to reproduce it in an identical environment. You can also try to use more MPI processes, which might probably trigger the error.

**Comment by hongriTianqi** [comment]
> I'm confused about why case `121_PW_KPAR` timed out ... I can't reproduce this problem on my workstation. @caic99

To me, it's total timeout, rather than 121_PW_KPAR timeout. 101_PW_15_lowz, for example, failed before 121.

**Comment by maki49** [comment]
> > I'm confused about why case `121_PW_KPAR` timed out ... I can't reproduce this problem on my workstation. @caic99
> 
> Hi @maki49 , I recommend to use a docker image to reproduce it in an identical environment. You can also try to use more MPI processes, which might probably trigger the error.

I reproduced the timeout of `121_PW_KPAR` on Bohrium. It occured (SCF not converged) only when ABACUS compiled with GNU and `OMP_NUM_THREADS=2`. But I have no knowledge about openmp ...

**Comment by maki49** [comment]
> > I'm confused about why case `121_PW_KPAR` timed out ... I can't reproduce this problem on my workstation. @caic99
> 
> Hi @maki49 , I recommend to use a docker image to reproduce it in an identical environment. You can also try to use more MPI processes, which might probably trigger the error.

I reproduced the timeout of `121_PW_KPAR on Bohrium. It occured  only when ABACUS compiled with GNU and `OMP_NUM_THREADS=2`. But I have no knowledge about openmp ...

**Comment by caic99** [comment]
> I reproduced the timeout of `121_PW_KPAR` on Bohrium. It occured (SCF not converged) only when ABACUS compiled with GNU and `OMP_NUM_THREADS=2`. But I have no knowledge about openmp ...

@maki49 Thanks! One more question: can you reproduce the error on the main branch (but not on this PR)?

**Comment by maki49** [comment]
> @maki49 Thanks! One more question: can you reproduce the error on the main branch (but not on this PR)?

Yes,  `deepmodeling::develop` branch also gives an ETOT larger than 1e+10 eV ...

image

**Comment by maki49** [comment]
> @maki49 Thanks! One more question: can you reproduce the error on the main branch (but not on this PR)?
Yes,  `deepmodeling::develop` branch also gives an ETOT larger than 1e+10 eV ...

image

**Comment by caic99** [comment]
> Yes, `deepmodeling::develop` branch also gives an ETOT larger than 1e+10 eV ...

Hi @maki49 ,
I can not reproduce your error.
Have you tried using a different number of threads?

**Comment by maki49** [comment]
> Hi @maki49 , I can not reproduce your error. Have you tried using a different number of threads?

4 cores, 4 threads: it got worse. 
image

4 cores, 1 thread: correctly converged.
image

**Comment by caic99** [comment]
> 4 cores, 4 threads: it got worse. image
> 
> 4 cores, 1 thread: correctly converged. image

@maki49 How many CPU cores do you have?

**Comment by maki49** [comment]
> @maki49 How many CPU cores do you have?

A node with c2_m4_cpu but it shows `nproc=16`. `mpirun -np 2 abacus` with `export OMP_NUM_THREADS=2` also reproduced this error.

**Comment by caic99** [comment]
> > @maki49 How many CPU cores do you have?
> 
> A node with c2_m4_cpu but it shows `nproc=16`. `mpirun -np 2 abacus` with `export OMP_NUM_THREADS=2` also reproduced this error.

@maki49 This is a known issue of Bohrium. You still have only 2 CPUs to use.
For ABACUS, you should not use OpenMP threads more than your assigned hardware, otherwise the result will be incorrect.

**Comment by maki49** [comment]
> @maki49 This is a known issue of Bohrium. You still have only 2 CPUs to use. For ABACUS, you should not use OpenMP threads more than your assigned hardware, otherwise the result will be incorrect.

I retried on a c16_m64 node using `mpirun -np 4 abacus` and `export OMP_NUM_THREADS=2`. Both `deepmodeling::develop` and `maki49::develop` have this problem, but `abacus-v3.1.0` converged correctly.

**Comment by maki49** [comment]
I've found the bug. Thanks for the help of @dyzheng , @caic99 , @hongriTianqi and @Liu-RX .
In function `rhog_piece_to_all`, `MPI_Send(tag=local_index)` got stuck because of a mismatch with `MPI_Recv(tag=global_index)`.
Now I've fixed it using local-to-local index .

Another change is to do `rhog_symmetry` respectively in each pool (which solved the problem that 3 and 4-cores' run give different results). 

The only thing left is to deal with deviations with result.ref, after which will I turn this PR "Ready for review".

**Review by dyzheng — APPROVED** [review]

**Review by hongriTianqi — APPROVED** [review]

**Review by pxlxingliang — APPROVED** [review]

File: source/module_cell/module_symmetry/symmetry_basic.h [hunk 1]
```diff
@@ -37,15 +37,17 @@ class Symmetry_Basic
...
-
+	void matrigen(ModuleBase::Matrix3 *symgen, const int ngen, ModuleBase::Matrix3* symop, int &nop) const;
+	void setgroup(ModuleBase::Matrix3 *symop, int &nop, const int &ibrav) const;
+	void rotate(
+			ModuleBase::Matrix3 &gmatrix, ModuleBase::Vector3<double> &gtrans, 
+			int i, int j, int k, const int, const int, const int, int&, int&, int&);
 	protected:
 
-	std::string get_brav_name(const int ibrav);
+	std::string get_brav_name(const int ibrav) const;
```
Comment (hongriTianqi) [inline source/module_cell/module_symmetry/symmetry_basic.h:49]: Could you please add unit tests for new added functions?

File: source/module_cell/module_symmetry/symmetry_basic.h [hunk 2]
```diff
@@ -37,15 +37,17 @@ class Symmetry_Basic
...
 	void pointgroup(const int &nrot,int &pgnumber,std::string &pgname, const ModuleBase::Matrix3* gmatrix, std::ofstream &ofs_running);
 	void atom_ordering(double *posi, const int natom, int *subindex);
+	void atom_ordering_new(double *posi, const int natom, int *subindex);
```
Comment (hongriTianqi) [inline source/module_cell/module_symmetry/symmetry_basic.h]: Could you please add unit test for newly added functions?

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/module_cell/module_symmetry/symmetry_basic.h (hunk 2 above)
Comment (maki49) [reply source/module_cell/module_symmetry/symmetry_basic.h]: Done in the latest commit.

**Review by hongriTianqi — APPROVED** [review]

File: source/module_cell/module_symmetry/test/symmetry_test.cpp [hunk 3]
```diff
@@ -279,21 +292,199 @@ class SymmetryTest : public testing::Test
...
+    }
+    //output old_pos and new_pos
+    std::cout<<"random direct coords: "<<std::endl;
+    std::cout<<"iatom     x      y     z"<<std::endl;
+    for (int i=0;i<nat;++i)
+        std::cout<<i<<" "<<old_pos[3*i]<<" "<<old_pos[3*i+1]<<" "<<old_pos[3*i+2]<<std::endl;
+    std::cout<<"sorted direct coords: "<<std::endl;
+    std::cout<<"iatom     x      y     z"<<std::endl;
+    for (int i=0;i<nat;++i)
+        std::cout<<i<<" "<<new_pos[3*i]<<" "<<new_pos[3*i+1]<<" "<<new_pos[3*i+2]<<std::endl;
```
Comment (hongriTianqi) [inline source/module_cell/module_symmetry/test/symmetry_test.cpp:488]: Is it better to add EXPECT_LE  here?

Comment (maki49) [reply source/module_cell/module_symmetry/test/symmetry_test.cpp:488]: Done, good idea.

File: source/module_cell/module_symmetry/test/symmetry_test.cpp [hunk 4]
```diff
@@ -6,6 +6,8 @@
 
 #include "gtest/gtest.h"
 
+#define DOUBLETHRESHOLD 1e-8
+
 /************************************************
  *  unit test of class Symmetry
  ***********************************************/
```
Comment (hongriTianqi) [inline source/module_cell/module_symmetry/test/symmetry_test.cpp:24]: It is suggested to add here a brief explanation of the newly added unittest. Thanks.

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

**Review by hongriTianqi — APPROVED** [review]

**Comment by caic99** [comment]
@maki49 Don't forget to close corresponding issues.

**Pull request merged**